pytensor/link/numba/dispatch/basic.py

```diff
@@ -280,104 +275,277 @@ def numba_funcify(op, node=None, storage_map=None, **kwargs):
     return generate_fallback_impl(op, node, storage_map, **kwargs)
 
 
-@numba_funcify.register(FunctionGraph)
-def numba_funcify_FunctionGraph(
-    fgraph,
-    node=None,
-    fgraph_name="numba_funcified_fgraph",
-    **kwargs,
-):
-    return fgraph_to_python(
-        fgraph,
-        numba_funcify,
-        type_conversion_fn=numba_typify,
-        fgraph_name=fgraph_name,
-        **kwargs,
-    )
+@singledispatch
+def numba_funcify_default_op_cache_key(
+    op, node=None, **kwargs
+) -> Callable | tuple[Callable, int]:
+    """Funcify an Op and allow a default cache key to be generated for it.
 
+    Wrapped function can return an integer in addition to the generated numba function.
 
-@numba_funcify.register(OpFromGraph)
-def numba_funcify_OpFromGraph(op, node=None, **kwargs):
-    _ = kwargs.pop("storage_map", None)
+    See docstrings of `register_funcify_default_op_cache_key` for details.
+    """
+    raise NotImplementedError()
 
-    # Apply inner rewrites
-    # TODO: Not sure this is the right place to do this, should we have a rewrite that
-    #  explicitly triggers the optimization of the inner graphs of OpFromGraph?
-    #  The C-code defers it to the make_thunk phase
-    fgraph = op.fgraph
-    add_supervisor_to_fgraph(
-        fgraph=fgraph,
-        input_specs=[In(x, borrow=True, mutable=False) for x in fgraph.inputs],
-        accept_inplace=True,
-    )
-    NUMBA.optimizer(fgraph)
-    fgraph_fn = numba_njit(numba_funcify(op.fgraph, **kwargs))
 
-    if len(op.fgraph.outputs) == 1:
+def register_funcify_default_op_cache_key(op_type):
+    """Funcify an Op and allow a default cache key to be generated for it.
 
-        @numba_njit
-        def opfromgraph(*inputs):
-            return fgraph_fn(*inputs)[0]
+    This function is a helper that dispatches to both `numba_funcify_default_op_cache_key`
+    and the legacy `numba_funcify`.
 
-    else:
+    The cache key will ultimately be generated by the base case of `numba_funcify_and_cache_key`
+    when a more specialized dispatch for the Op is not registered. Function wrapped by this decorator
+    can return an integer in addition to the numba function.
+    This will be added to the default cache key, and can be used to signal changes over versions.
 
-        @numba_njit
-        def opfromgraph(*inputs):
-            return fgraph_fn(*inputs)
+    The default cache key is based on the string representations of: `type(op)` and the
```

Comment:
Maybe we should also add the pytensor version to this?
If we don't, I think we have to always increase the counter for any implementation change from now on, and I'm a bit scared of the kind of bugs that this could lead to if we ever forget or mess it up somehow. (Only reproducible of you run the old version of pytensor first...).
This would slow down compilation after every pytensor update, but I think that might be worth the cost.
This would leave the counter for development.